src/dynamic_programming/knapsack.md

```diff
@@ -0,0 +1,157 @@
+(In fact, this is exactly the solution to the complete knapsack problem)
+
+To avoid this, we can change the order of enumeration from $W$ to $w_{i}$, so that the above error won't occour, because $f_{i, j}$ is always updated before $f_{i, j-w_i}$
+
+Therefore, the actual code is
+
+```.c++
+for (int i = 1; i <= n; i++)
+  for (int j = W; j >= w[i]; j--)
+    f[j] = max(f[j], f[j - w[i]] + v[i]);
+```
+
+## Complete Knapsack
+
+The complete knapsack model is similar to the 0-1 knapsack, the only difference from the 0-1 knapsack is that an item can be selected an unlimited number of times instead of only once.
+
+We can refer to the idea of 0-1 knapsack to define the state: $f_{i, j}$, the maximum value the knapsack can obtain using the first $i$ items with maximum capacity $j$.
+
+It should be noted that although the state definition is similar to that of a 0-1 knapsack, its state transfer equation is different from that of a 0-1 knapsack.
+
+### Explaination
+
+The trivial approach is, for the first $i$ items, enumerate how many each item is to be taken. The time complexity of this is $O(n^3)$.
+
+The state transfer function follows:
+
+$$f_{i, j} = \max\limits_{k=0}^{\infty}(f_{i-1, j-k\cdot w_i} + k\cdot v_i)$$
+
+It can be found that for $f_{i, j}$, it can just transfer through $f_{i, j-w_i}$. The state transfer equation becomes:
+
+$$f_{i, j} = \max(f_{i-1, j},f_{i, j-w_i} + v_i)$$\
+
+The reason this works is that when we transfer like this, $f_{i, j-w_i}$ has already been updated by $f_{i, j-2\cdot w_i}$ and so on.
+
+Similar to the 0-1 knapsack, we can remove the first dimension to optimize the space complexity.
+
+## Multiple Knapsack
+
+Multiple knapsack is also a variant of 0-1 knapsack. The main difference is that there are $k_i$ of each item instead of just 1.
+
+### Explaination
+
+A very simple idea is: "choose each item $k_i$ times" is equivalent to "$k_i$ of the same item is selected one by one". Thus converting it to a 0-1 knapsack model, which can be described by the transition function:
+
+$$f_{i, j} = \max_{k=0}^{k_i}(f_{i-1,j-k\cdot w_i} + k\cdot v_i)$$
+
+The time complexity of this process is $O(W\sum\limits_{i=1}^{n}k_i)$
+
+### Binary Grouping Optimization
+
+We still consider converting the multiple knapsack model into a 0-1 knapsack model for optimization. The time complexity $O(Wn)$ can not be further optimized, so we focus on $O(\sum k_i)$.
+
+Let $A_{i, j}$ denote the $j^{th}$ item split from the $i^{th}$ item. In the trivial approach discussed above, $A_{i, j}$ represents the same item for all $j \leq k_i$. The main reason for our low efficiency is that we are doing a lot of repetetive work. For example, consider selecting $A_{i, 1} and A_{i, 2}$, and selecting $A_{i, 3} and A_{i, 3}$. These two situations are completely equivalent. Thus optimizing the spiltting method will greatly reduces the time complexity.
+
+The grouping is made more effiecent by using binary grouping.
```

Comment:
This is an interesting idea, but it will increase the number of items by a factor of $\log k_i$ making it $O(W n \log k)$. I think true $O(Wn)$ multiple knapsack requires maintaining a [maximum queue](https://cp-algorithms.com/data_structures/stack_queue_modification.html) of size $k_i$ for each remainder modulo $w_i$, and these nuances should be made more explicit.

Reply:
I'm not familiar with the method you mentioned. Indeed the method I've mentioned is $O(W \sum \log k_i) \approx O(Wn\log k)$, which seems good enough for such problem. Eg: [CSES - Book Shop II](https://cses.fi/problemset/task/1159/), [sample solution](https://pastebin.com/7tXAaF6r)

Reply:
The solution I'm describing is, assume you're currently processing an item described by $(w_i, v_i, k_i)$.

We have the transition rule $f_j \gets \max\limits_{k=0..k_i}(f_{j-kw_i} + kv_i) = j v_i + \max\limits_{k=0..k_i}g_{j-kw_i}$, where $g_{j} = f_j - j v_i$.

Now, for each remainder $j \bmod w_i$, you may independently store a maximum queue, so that you can extract the largest number among the last $k_i$ elements in $O(1)$ amortized. This way, we get a solution with total runtime of pure $O(nW)$, without $\log k$ factor.